### DISCUSSION TITLE
Fabric JS + Redux (Undo/Redo)

### DISCUSSION THREAD

**Discussion by vavilon224** [fabricjs/fabric.js#9702] [Q&A] [answered]
Hello everyone, I need help saving "canvas.toDatalessJSON(HISTORY_IMPORT_PARAMS);"
To create undo redo it lags a lot, what can you advise and how to implement it if I have a very large amount of data??

```
updateHistory(state, action) {
      const { past, present, currentSide, canvasIsReadyForUpdate } = state;
      const response = action.payload;

      if (present && currentSide && !canvasIsReadyForUpdate) {
        const pastBySide = past[currentSide] || [];
        const newPast = [...pastBySide, present[state.currentSide]];

        if (newPast.length > 15) {
          newPast.splice(0, newPast.length - 15);
        }

        state.canvasIsReadyForUpdate = false;
        state.past = {
          ...past,
          [currentSide]: newPast,
        };
        state.present = {
          ...present,
          [currentSide]: response,
        };
        state.future = {};
      }
    },
```

@ShaMan123 please help

**Comment by vavilon224** [comment]
The point is that I have sides of which there can be 5, and every time I change I save a huge array of objects in redux.
Accordingly, the browser cannot handle it, what should I do?

**Reply by vavilon224** [reply]
@ShaMan123 please help

**Comment by vavilon224** [comment]
The data flow is very large and where to save this data is not clear; sometimes it reaches 10 MB
but since I also have 15 undo 15 redo.
150 MB of data where and how to save it in real time

**Comment by ShaMan123** [comment, marked as answer]
I can't really help without a repro and also with a repro because it is deeply connected to your app's business logic.
Mind that it is very wasteful to save the entire canvas data each history step.
Using `object:modified` event for example and other triggers can help you create an object based history.
I like to inspire from the diffing master `git`: save the minimal diff of each step and have a way to apply that diff back and forth. That way jumping through history is as a simple as a loop of diff application and the size of it is very small.
Choose a diff tool or handle all the cases yourself if that makes sense to produce your own diff e.g.
```
// object diff
[objectId, { fill: 'red' }]
[objectId2, { width: 10, height: 10 }]
[objectId, { stackIndex: 4 }]
[objectId, { parent: parentId, stackIndex: 2 }]
...
```
Think what is best for you

**Reply by ShaMan123** [reply]
Note that if you have grouping operations it is wise to save the object's position in the scene plane using `sendPointToPlane` or `object.getCenterPoint()`

**Reply by ShaMan123** [reply]
If you choose a very good diffing solution then you could diff the entire canvas json output and recreate everything each history step. However if you have a huge object tree that will damage performance

**Comment by asturur** [comment]
Another approach for undo/redo is to use actions.
Since you have redux you probably dispatch action.
So for example when someone change the color of an object from red to green, you can dispatch an action that says:
- change: object with id abc becomes green
- undoChange: ( here you put the whole action that makes you object turn back red )

change will mutate the state while undoChange gets stored to be replayed later.

Every large action needs to  be broken in steps your code can understand.
But as @ShaMan123  said it all depends on how your app is setup.
